src/trace-plugin-loader.ts

```diff
@@ -15,26 +15,113 @@
  */
 'use strict';
 
+// TODO(kjin)
+// Module augmentation is implemented in this PR:
+// https://github.com/DefinitelyTyped/DefinitelyTyped/pull/19612
+// Do this correctly when it's landed and released.
+import _Module = require('module');
+const Module: {
+  _resolveFilename(request: string, parent?: NodeModule): string;
+  _load(request: string, parent?: NodeModule, isMain?: boolean): any;
+} = _Module as any;
+
+import { Logger } from '@google-cloud/common';
+import * as path from 'path';
+import * as semver from 'semver';
+import * as shimmer from 'shimmer';
 import * as util from './util';
 
-var Module = require('module');
-var shimmer = require('shimmer');
-var path = require('path');
-var semver = require('semver');
-var TraceAgent = require('./trace-api'/*.js*/);
+const TraceAgent = require('./trace-api');
+// TODO(kjin): When it's is converted, import these from trace-api
+export interface TraceAgentConfig {}
+export interface TraceAgent {
+  enable: (logger: Logger, config: TraceAgentConfig) => void;
+  disable: () => void;
+}
+
+/**
+ * An interface representing config options read by the plugin loader, which includes
+ * TraceAgent configuration as well.
+ */
+export interface PluginLoaderConfig extends TraceAgentConfig {
+  plugins: {
+    [pluginName: string]: string;
+  };
+}
+interface InternalPatch<T> {
+  file: string;
+  module?: T;
+  patch: (module: T, agent: TraceAgent) => void;
+  unpatch?: (module: T) => void;
+}
+
+interface InternalIntercept<T> {
+  file: string;
+  module?: T;
+  intercept: <T>(module: T, agent: TraceAgent) => T;
+}
+
+interface InternalPlugin {
+  file: string;
+  patches: {
+    [patchName: string]: {
+      [file: string]: InternalPatch<any> | InternalIntercept<any>;
+    }
+  };
+  agent: TraceAgent;
+}
+
+interface PluginStore {
+  [pluginName: string]: InternalPlugin;
+}
+
+export interface Patch {
+  file?: string;
+  versions?: string;
+  patch: (module: any, agent: TraceAgent) => void;
+  unpatch?: (module: any) => void;
+}
+
+export interface Intercept {
+  file?: string;
+  versions?: string;
+  intercept: <T>(module: T, agent: TraceAgent) => T;
+}
+
+export type Plugin = (Patch | Intercept)[];
```

Comment:
style guide prefers `Array<>` syntax.

Reply:
it might make sense to give a concrete type name to `Patch|Intercept`.

Reply:
`Array<>` syntax done.

Thoughts on `type Instrumentation = Patch | Intercept`? I realize the term is overloaded in the current file, but the variables named `instrumentation` are all internal and can be renamed.